### DISCUSSION TITLE
Reakcja funkcji phfwdGet na niepoprawne dane

### DISCUSSION THREAD

**Discussion by Alef-Zero** [Kwasow/IPP_testy#25] [General]
Zgodnie ze specyfikacją funkcja phfwdGet powinna zwracać NULL w sytuacji,  w której nie udało się zaaolokować pamięci

 * @return Wskaźnik na strukturę przechowującą ciąg numerów lub **NULL, gdy nie udało się alokować pamięci.**

moim zdaniem w sytuacji, w której funkcja na wejściu dostanie dane niepoprawne powinna zwrócić wskaźnik do struktury PhoneNumbers takiej, że odpytanie o wartośc pod indeksem 0 zwróci pusty ciąg "" a odpytanie o kolejne indeksy zwróci NULL.

Implementacja funkcji phfwdGet w taki sposób, żezwraca NULL w sytuacji otrzymania niepoprawnychj danych wejściowych wydaje się być nadinterpretacją.

Co sądzicie na ten temat?

**Comment by Kwasow** [comment]
Jest to w sumie sensowny sposób patrzenia na to, o którym nie pomyślałem. Wydaje mi się, że to pytanie jak najbardziej nadaje się na forum i jest bezpieczne, bo to jest kwestia jednego ifa.

Co sądzicie @Rentib, @pggPL ?

**Comment by pggPL** [comment]
No można zadać, ale zapytałbym się na konfie roku najpierw, żeby nikt nie miał pretensji potem.
Rozumiem, że chodzi o to, że phfwdGet zwraca NULL już dla 0?

**Comment by Kwasow** [comment]
W naszych testach jest tak, że jeżeli numer jest niepoprawny w phfwdGet, to phfwdGet zwraca NULL. Tymczasem propozycja jest, żeby zwracał wtedy pustą strukturę PhoneNumbers i wtedy phnumGet na tej strukturze zwracałoby NULL dla każdego indeksu

**Comment by Alef-Zero** [comment]
Ponieważ tak naprawdę chodzi o to, żeby obsłużyć sytuacje błędnych  danych, przynajmniej do czasu rozstrzygnięcia "u źródla" jak powinno być, programy testujące powinny dopuszczać obie możliwości. Na przykład:

  _pnum = phfwdGet(NULL, "123");
  assert((pnum == NULL) || (strcmp(phnumGet(pnum, 0), "") == 0 ));
  if( pnum != NULL ) phnumDelete(pnum);_

**Reply by Kwasow** [reply]
Zgodnie z informacją, którą ktoś podał na grupie, powinna być pusta struktura (Peczarski mówił na labach).

Zmienię dzisiaj, dzięki za zwrócenie na to uwagi.

**Comment by Kwasow** [comment]
Krótki update. Jeśli podany numer jest niepoprawny w `phfwdGet`, to funkcja zwraca strukturę PhoneNumbers, która nie jest NULLem, ale nie ma żadnych numerów, tzn. dla każdego indeksu w pnumGet zwraca NULL.
